src/protocols/message_counter/MessageCounterManager.cpp

```diff
@@ -194,97 +240,88 @@ CHIP_ERROR MessageCounterSyncMgr::AddToReceiveTable(const PacketHeader & packetH
 
             // Explicitly free any buffer owned by this handle.  The
             // HandleGroupMessageReceived() call should really handle this, but
             // just in case it messes up we don't want to get confused about
             // wheter the entry is in use.
-            entry.msgBuf = nullptr;
+            entry.msgBuf     = nullptr;
+            entry.peerNodeId = kUndefinedNodeId;
         }
     }
 }
 
 // Create and initialize new exchange for the message counter synchronization request/response messages.
-CHIP_ERROR MessageCounterSyncMgr::NewMsgCounterSyncExchange(SecureSessionHandle session,
+CHIP_ERROR MessageCounterManager::NewMsgCounterSyncExchange(SecureSessionHandle session,
                                                             Messaging::ExchangeContext *& exchangeContext)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
-
-    // Message counter synchronization protocol is only applicable for application group keys.
-    VerifyOrReturnError(ChipKeyId::IsAppGroupKey(session.GetPeerKeyId()), err = CHIP_ERROR_INVALID_ARGUMENT);
-
     // Create new exchange context.
     exchangeContext = mExchangeMgr->NewContext(session, this);
-    VerifyOrReturnError(exchangeContext != nullptr, err = CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(exchangeContext != nullptr, CHIP_ERROR_NO_MEMORY);
 
-    return err;
+    return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR MessageCounterSyncMgr::SendMsgCounterSyncReq(SecureSessionHandle session)
+CHIP_ERROR MessageCounterManager::SendMsgCounterSyncReq(SecureSessionHandle session, Transport::PeerConnectionState * state)
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
     Messaging::ExchangeContext * exchangeContext = nullptr;
-    Transport::PeerConnectionState * state       = nullptr;
     System::PacketBufferHandle msgBuf;
     Messaging::SendFlags sendFlags;
-    uint8_t challenge[kMsgCounterChallengeSize];
-
-    state = mExchangeMgr->GetSessionMgr()->GetPeerConnectionState(session);
-    VerifyOrExit(state != nullptr, err = CHIP_ERROR_NOT_CONNECTED);
+    std::array<uint8_t, kChallengeSize> challenge;
 
     // Create and initialize new exchange.
     err = NewMsgCounterSyncExchange(session, exchangeContext);
     SuccessOrExit(err);
 
     // Allocate a buffer for the null message.
-    msgBuf = MessagePacketBuffer::New(kMsgCounterChallengeSize);
+    msgBuf = MessagePacketBuffer::New(kChallengeSize + kMacSize);
     VerifyOrExit(!msgBuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
 
     // Generate a 64-bit random number to uniquely identify the request.
-    err = DRBG_get_bytes(challenge, kMsgCounterChallengeSize);
+    err = DRBG_get_bytes(challenge.data(), kChallengeSize);
     SuccessOrExit(err);
 
     // Store generated Challenge value to ExchangeContext to resolve synchronization response.
-    exchangeContext->SetChallenge(challenge);
+    state->GetSessionMessageCounter().GetPeerMessageCounter().StartSync(challenge);
```

Comment:
Does the challenge value stored into ExchangeContext? it seems not

Reply:
It is stored inside `PeerMessageCounter`, I'll remove the comment